resolwe/flow/tests/test_purge.py

```diff
@@ -0,0 +1,223 @@
+
+re-save-file sample_file this_file_should_stay ref3
+re-save-dir sample_dir stay ref4
+re-save-file-list sample_file_list entry1 entry2:ref1 entry3:refs
+re-save-dir-list sample_dir_list dir1:ref2 dir2 dir3
+""",
+                }
+            )
+        )
+
+        self.assertFilesRemoved(data, 'these', 'files', 'should', 'be', 'removed', 'directory')
+        self.assertFilesNotRemoved(data, 'this_file_should_stay', 'stay/directory/file1', 'stay/directory/file2',
+                                   'entry1', 'entry2', 'entry3', 'dir1/a', 'dir2/b', 'dir3/c', 'dir3/d', 'ref1',
+                                   'ref2', 'ref3', 'ref4', 'refs/a', 'refs/b')
+
+    def assertFieldWorks(self, field_type, field_value, script_setup, script_save, removed, not_removed):
+        field_schema = {
+            'name': 'sample',
+            'label': 'Sample output',
+            'type': field_type
+        }
+
+        # Test simulated operation.
+        simulated_root = tempfile.mkdtemp()
+        try:
+            for filename in removed + not_removed:
+                directory, basename = os.path.split(filename)
+                if directory:
+                    try:
+                        os.makedirs(os.path.join(simulated_root, directory))
+                    except OSError:
+                        pass
+
+                if basename:
+                    with open(os.path.join(simulated_root, filename), 'w'):
+                        pass
+
+            unreferenced_files = purge.get_purge_files(
+                simulated_root,
+                output={'sample': field_value},
+                output_schema=[field_schema],
+                descriptor={},
+                descriptor_schema=[]
+            )
+
+            for filename in not_removed:
+                self.assertNotIn(os.path.realpath(os.path.join(simulated_root, filename)), unreferenced_files)
+
+            for filename in removed:
+                filename = os.path.realpath(os.path.join(simulated_root, filename))
+                self.assertIn(filename, unreferenced_files)
+                unreferenced_files.discard(filename)
+
+            # Ensure that nothing more is removed.
+            self.assertEqual(len(unreferenced_files), 0)
+        finally:
+            shutil.rmtree(simulated_root)
+
+        # Test output.
+        data = self.create_and_run_processor(
+            processor=dict(
+                input_schema=[],
+                output_schema=[field_schema],
+                run={
+                    'bash': script_setup + '\n' + script_save
+                }
+            )
+        )
+
+        self.assertFilesRemoved(data, *removed)
+        self.assertFilesNotRemoved(data, *not_removed)
+
+        # Test descriptor.
+        descriptor_schema = DescriptorSchema.objects.create(
+            slug=get_random_string(6),
+            contributor=self.admin,
+            schema=[field_schema]
+        )
+        data = self.create_and_run_processor(
```

Comment:
The `assertFieldWorks` checks multiple things, but only for a specific field type. And then `test_general` contains a more complex example with multiple fields on the same processor.

I am not sure why do you think the first part is more complex? I wanted to make sure that when you call `assertFieldWorks`, this tests everything for that field. And the different checks that are performed (test simulated operation, test output and test descriptor) are documented in the function. I don't want to see that you need to write two different test functions (or more) for each field, because this means that it is harder to extend and some test cases can be missed when adding a new field.

This makes it easy to extend test cases:
* When adding a new field, one just needs to call `self.assertFieldWorks` with the correct parameters.
* When extending tests for each field, one just needs to modify the `assertFieldWorks` to include additional checks for each field, without the need to modify all the tests.